src/aof.c

```diff
@@ -40,189 +40,692 @@
+     * we don't need atomic modification here. */
+    sds aof_name = getLastIncrAofName(server.aof_manifest);
+    /* Here we should use 'O_APPEND' flag. */
+    server.aof_fd = open(aof_name, O_WRONLY|O_APPEND|O_CREAT, 0644);
+    if (server.aof_fd == -1) {
+        serverLog(LL_WARNING, "Can't open the append-only file %s: %s", 
+            aof_name, strerror(errno));
+        exit(1);
+    }
+
+    int ret = persistAofManifest(server.aof_manifest);
+    if (ret != C_OK) {
+        exit(1);
+    }
+
+    server.aof_last_incr_size = getAppendOnlyFileSize(aof_name);
+}
+
+/* Called in `rewriteAppendOnlyFileBackground`. If `server.aof_state` 
+ * is 'AOF_ON' or ‘AOF_WAIT_REWRITE', It will do two things:
+ * 1. Open a new INCR type AOF for writing
+ * 2. Synchronously update the manifest file to the disk
+ * 
+ * The above two steps of modification are atomic, that is, if 
+ * any step fails, the entire operation will rollback and returns
+ * C_ERR, and if all succeeds, it returns C_OK.
+ * */
+int openNewIncrAofForAppend(void) {
+    serverAssert(server.aof_manifest != NULL);
+    int newfd;
+
+    /* Only open new INCR AOF when AOF enabled. */
+    if (server.aof_state == AOF_OFF) return C_OK;
+
+    /* Dup an temp aof_manifest to modify. */
+    aofManifest *temp_am = aofManifestDup(server.aof_manifest);
+    if (temp_am == NULL) return C_ERR;
+
+    /* Open new AOF. */
+    sds new_aof_name = getNewIncrAofName(temp_am);
+    newfd = open(new_aof_name, O_WRONLY|O_TRUNC|O_CREAT, 0644);
+    if (newfd == -1) {
+        serverLog(LL_WARNING, "Can't open the append-only file %s: %s",
+            new_aof_name, strerror(errno));
+
+        aofManifestFree(temp_am);
+        return C_ERR;
+    }
+
+    /* Persist AOF Manifest. */
+    int ret = persistAofManifest(temp_am);
+    if (ret == C_ERR) {
+        close(newfd);
+        aofManifestFree(temp_am);
+        return C_ERR;
+    }
+
+    /* If reaches here, we can safely modify the `server.aof_manifest` 
+     * and `server.aof_fd`. */
+
+    /* Close old aof_fd if needed. */
+    if (server.aof_fd != -1) bioCreateCloseJob(server.aof_fd);
+    server.aof_fd = newfd;
+
+    /* Reset the aof_last_incr_size. */
+    server.aof_last_incr_size = 0;
+    /* Update `server.aof_manifest`. */
+    aofManifestFreeAndUpdate(temp_am);
+    return C_OK;  
+}
+
+/* Whether to limit the execution of Background AOF rewrite.
+ * 
+ * At present, if AOFRW fails, redis will automatically retry. If it continues 
+ * to fail, we may get a lot of very small INCR files. so we need an AOFRW 
+ * limiting measure.
+ * 
+ * We can't directly use `server.aof_current_size` and `server.aof_last_incr_size`, 
+ * because there may be no new writes after AOFRW fails.
+ * 
+ * So, we use time delay to achieve our goal. When AOFRW fails, we delay the execution 
+ * of the next AOFRW by 1 minute. If the next AOFRW also fails, it will be delayed by 2 
+ * minutes. The next is 4, 8, 16, the maximum delay is 60 minutes (1 hour).
```

Comment:
LOL.. i was arguing for a day.
i think there are cases where something can fail for weeks and no one will notice.
as discussed in another thread, if the admin solves the problem he can run BGREWRITEAOF instead of waiting for the next interval, so why make it that short?

Reply:
Yes, I discussed with @oranagra . It was originally one day. I changed it to 1 hour later. In any case, once the user finds and solves the related problem, he can use BGREWRITEAOF to execute AOFRW immediately without waiting for the limit. So I think there is no need to change here. @yossigo WDYT?